pysixdesk/lib/study.py

```diff
@@ -692,27 +689,29 @@ def submit(self, typ, trials=5, *args, **kwargs):
 
         batch_name = os.path.join(self.study_path, jobname)
         where = "batch_name like '%s_%%'" % batch_name
-        que_out = self.db.select(
-            table_name, 'batch_name', where, DISTINCT=True)
+        que_out = self.db.select(table_name, 'batch_name',
+                                 where, DISTINCT=True)
         ibatch = len(que_out)
         ibatch += 1
         batch_name = batch_name + '_' + str(ibatch)
-        status, out = self.submission.submit(input_path, batch_name,
-                                             trials, *args, **kwargs)
-        if status:
-            content = "Submit %s job successfully!" % jobname
-            self._logger.info(content)
-            table = {}
-            table['status'] = 'submitted'
-            for ky, vl in out.items():
-                where = 'wu_id=%s' % ky
-                table['unique_id'] = vl
-                table['batch_name'] = batch_name
-                self.db.update(table_name, table, where)
-        else:
-            # self.purge_table(task_table_name)
+
+        try:
+            out = self.submission.submit(input_path, batch_name,
+                                         trials, *args, **kwargs)
+        except Exception as e:
```

Comment:
I think try-except is not suitable here. 
I used a script to do testing
```python
a=WorkSpace()
test=a.init_study('local')
test=a.load_study('local')
test.prepare_preprocess_input()
test.submit(0,5)
test.collect_result(0)
```
If I run this script twice, it will complaint `FileNotFoundError`, but it's actually not an error, just tell you there isn't unsubmitted job, and the script should go on to collect the results. But If we remove `raise FileNotFoulndError` statement here, it will lead to another problems. A `status` is better here. 
So sometimes we should avoid to use `raise` statement, if the 'error' doesn't influence next step, the program shouldn't be stopped.

Reply:
Yes, good catch ! I still don't think using a status variable is necessary. You can, for example, wrap the `submit` call in your script with a try/except to suppress the exception.

This allows you to know during script execution if the submit failed and act on it, this would not be possible with status variables unless you carry them all the way up, which is not scalable and would get very clunky very fast. 

If you don't like the idea of wrapping the `submit` call with try/except, then you can just switch the `raise` for a `return`  as shown bellow and it should work ? Again, doing this won't allow any scripting on top of pysixdesk to know if the submission failed, because the exception is not raised.

I would prefer to leave the raising of the exception and encourage the use of try/except (the same way the built-in functions, e.g.`open` raises a FileNotFoundError instead of printing the error msg and continuing) as it is more flexible and gives more options to the user if they want to write scripts on top of pysixdesk.

```patch
diff --git a/pysixdesk/lib/study.py b/pysixdesk/lib/study.py
index 3c62d30..7a64ce8 100644
--- a/pysixdesk/lib/study.py
+++ b/pysixdesk/lib/study.py
@@ -699,10 +699,10 @@ class Study(object):
         try:
             out = self.submission.submit(input_path, batch_name,
                                          trials, *args, **kwargs)
-        except Exception as e:
+        except Exception:
             content = "Failed to submit %s job!" % jobname
             self._logger.error(content)
-            raise e
+            return
 
         content = "Submit %s job successfully!" % jobname
         self._logger.info(content)
```

Let me know what you think !

Reply:
I agree with you, using `try-except` is a good  habit. 
But, wrap the `submit` call in the user's script is not a wise choice, the user should not take care of this.
And I only thought it's not suitable here. Within `submit` method, there is already `try-except` statment to catch some errors. If we just replace `raise e`  with `return`, it will miss the real errors.
I think the better way is keepping the `try-except` within the `submit` method,  for the real serious we shoud use `raise` statement to stop the program, but for others we only need to print a warning message.

Reply:
> But, wrap the `submit` call in the user's script is not a wise choice, the user should not take care of this.

As I mentioned before, I think the increase in scripting flexibility of raising the exceptions, outweighs expecting the user to use try/except, especially seeing as it conforms to python's built-ins and ecosystem.

> And I only thought it's not suitable here. Within `submit` method, there is already `try-except` statment to catch some errors. If we just replace `raise e` with `return`, it will miss the real errors.

Yes you are right, how about this ? This should suppress the FileNotFoundError and the potential error during submission, but would not suppress any other exceptions.

```patch
diff --git a/pysixdesk/lib/study.py b/pysixdesk/lib/study.py
index 3c62d30..f2dc74a 100644
--- a/pysixdesk/lib/study.py
+++ b/pysixdesk/lib/study.py
@@ -699,10 +699,10 @@ class Study(object):
         try:
             out = self.submission.submit(input_path, batch_name,
                                          trials, *args, **kwargs)
-        except Exception as e:
+        except FileNotFoundError, submission.SubmissionError:
             content = "Failed to submit %s job!" % jobname
             self._logger.error(content)
-            raise e
+            return
 
         content = "Submit %s job successfully!" % jobname
         self._logger.info(content)
diff --git a/pysixdesk/lib/submission.py b/pysixdesk/lib/submission.py
index dd985a3..69e638a 100644
--- a/pysixdesk/lib/submission.py
+++ b/pysixdesk/lib/submission.py
@@ -8,6 +8,10 @@ from subprocess import Popen, PIPE
 from . import utils
 
 
+class SubmissionError(Exception):
+    pass
+
+
 class Cluster(ABC):
 
     def __init__(self, temp_path):
@@ -135,7 +139,7 @@ class HTCondor(Cluster):
                     uniq_ids = [str(cl_id) + '.' + str(pr_id) for pr_id in proc_ls]
                     if len(wu_ids) != len(uniq_ids):
                         content = "There is something wrong during submitting!"
-                        raise Exception(content)
+                        raise SubmissionError(content)
                     else:
                         comb = list(zip(wu_ids, uniq_ids))
                         out = dict(comb)

```

Reply:
We really don't need to create an custom exception, it doesn't make any sense.
We already have the simple way, there is  no need to do redundant things just for keep using `raise FileNotFoundError`. 
Let's just keep the original logic, this issue doesn't worth for further disscussion.

Reply:
Sometimes the simplest way isn't the correct way, but feel free to change it however you see fit.

Just quick note on the topic of custom exceptions, for your information, when I replaced status variables with exceptions, there were quite a few cases where I had to use the base `Exception` because there are no built in exceptions which cover the error, this is not good practice because the base `Exception` [covers most of the built-in exceptions](https://docs.python.org/3/library/exceptions.html#exception-hierarchy), leads to exactly this issue we were discussing here, where we cannot catch the desired exception without catching all the others. Custom exceptions **are** the correct solution to problems like this. Usually you put all your custom expression into a separate file and import them when needed. So at some point it would be a good idea to replace all the base `Exceptions` I used with custom exceptions, to avoid problems like these.